pandas/tests/io/parser/usecols.py

```diff
@@ -475,3 +476,46 @@ def test_uneven_length_cols(self):
                               'C': [3, 5, 4, 3, 3, 7]})
         df = self.read_csv(StringIO(data), usecols=usecols)
         tm.assert_frame_equal(df, expected)
+
+    def test_raise_on_usecols_names_mismatch(self):
+        ## see gh-14671
+        data = 'a,b,c,d\n1,2,3,4\n5,6,7,8'
+        msg = 'Usecols do not match names'  ## from parsers.py CParserWrapper()
```

Comment:
Condition the message on `self.engine` i.e.:
~~~
msg = <first-message> if self.engine == 'c' else <second-message>
~~~
That way you don't need that massive regex (and can remove the `re` import)